### ISSUE TITLE
False positive of protocol conforming when `Self` is used in a base protocol

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#4966]
**Describe the bug**
When a protocol is inheriting another protocol, and the base protocol uses `Self` in some way, conforming to the derived protocol sometimes fails unexpectedly.

**To Reproduce**
See sample code.

**Expected behavior**
Both assignments in the last should type-check.

**Screenshots or Code**
```python
from __future__ import annotations

from typing import Generic, Protocol, TypeVar
from typing_extensions import Self

_V_co = TypeVar("_V_co", covariant=True)

class P0(Protocol[_V_co]):
    def f0(self, /) -> Self: ...
    def f1(self, /) -> _V_co: ...

class P1(P0[_V_co], Protocol[_V_co]):
    pass

class C(Generic[_V_co]):
    def f0(self, /) -> Self:
        raise NotImplementedError
    def f1(self, /) -> _V_co:
        raise NotImplementedError

a: P0[str] = C[str]()  # pass
b: P1[str] = C[str]()  # fail
```

![image](https://user-images.githubusercontent.com/15389744/232662110-3fd296c3-62b0-477f-8b13-41df92d5a006.png)

**VS Code extension or command-line**
Pylance v2023.4.21 Pre-Release

**Additional context**
If `C` explicitly inherits `P1[_V_co]`, no error is reported.

Mypy playground: https://mypy-play.net/?mypy=latest&python=3.10&gist=a26280d31899e845ae73aecf9068bafa

EDIT
----
A derived class is not needed to repro.

**Comment by erictraut** [comment]
Thanks for the bug report and great repro steps.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.306, which I just published. It will also be included in a future release of pylance.

**Comment by Azureblade3808** [comment]
Still works incorrectly over some cases in Pylance v2023.5.11 (Pyright 1.1.306).

```python
from __future__ import annotations

from typing import Protocol, TypeVar
from typing_extensions import Self

_V_t = TypeVar("_V_t")
_V_co = TypeVar("_V_co")

class SetLike(Protocol[_V_co]):
    def __and__(self, right: Self, /) -> SetLike[_V_co]: ...
    # Other methods are omitted.

class MutableSetLike1(SetLike[_V_t], Protocol[_V_t]):
    def add(self, o: _V_t, /) -> None: ...
    # Other methods are omitted.
    
class MutableSetLike2(Protocol[_V_t]):
    def __and__(self, right: Self, /) -> SetLike[_V_t]: ...
    def add(self, o: _V_t, /) -> None: ...
    # Other methods are omitted.

SOME_SET: set[str] = {"A", "B", "C"}

a: SetLike[str] = SOME_SET
b: MutableSetLike1[str] = SOME_SET
c: MutableSetLike2[str] = SOME_SET
```
![image](https://github.com/microsoft/pyright/assets/15389744/67efc34a-27e0-4cb2-8ea9-ced12c9766a7)

In the sample above, `MutableSetLike1` and `MutableSetLike2` should work equally, but the results are different.

Mypy playground: https://mypy-play.net/?mypy=latest&python=3.10&gist=d68e3371c260e7b333d059194b1b9774

**Comment by erictraut** [comment]
@Azureblade3808, this issue is closed. If you want to report a new bug, please open a new issue.
